refactor: implement MVC architecture and fix critical bugs (#7, #8, #10, #11, #13, #14, #15, #16

itumelengseema opened the pull request:
- Restructured project to follow MVC pattern with SOLID principles
- Created services layer (ApiService, CacheService, AudioService, FirebaseService)
- Implemented Result<T> type for functional error handling
- Added custom exception hierarchy for type-safe errors
- Fixed email validation to accept all valid domains
- Fixed favorites isolation per user
- Fixed audio service singleton pattern
- Fixed API query parameter handling
- Added multi-layer caching strategy
- Prepared architecture for web compatibility

Closes #7, #8, #10, #11, #13, #14, #16
Fixed #6,#19 (architecture foundation)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * App-wide audio system with theme music and SFX.
  * Multi-layer caching for faster/offline Pokémon access.
  * Centralized app constants for configuration and assets.
  * New error/result abstractions for consistent operation outcomes.
  * Enhanced authentication with per-user data isolation and validation.
  * More robust Pokémon models with safer JSON parsing and value equality.

* **Bug Fixes**
  * Stronger email validation using a stricter regex.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

itumelengseema committed:
refector: implemented MVC And fixed critical Bugs

itumelengseema committed:
refector and fixed major bugs

itumelengseema committed:
Merge branch 'main' into refactor/mvc-architecture-and-bug-fixes

Joashp01 committed:
Update lib/models/pokemon.dart

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

Joashp01 committed:
Update test/models/user_test.dart

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

Joashp01 committed:
Update test/providers/theme_provider_test.dart

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

Joashp01 reviewed (changes requested):
# Bug Report: Pokedex Application Issues
1. **Pokemon Colors Not Correlating to Types** - Medium Severity
2. **Favorites Persisting Across Different Users** - High Severity 



---

## Issue #1: Pokemon Colors Not Correlating to Types

### Severity: **MEDIUM** 

### Description
 Pokemon colors no longer correlate with their types. All Pokemon appear to have the same gray color in the list view, rather than displaying colors that match their primary type (fire = red/orange, water = blue, etc.).

### Status
The color mapping logic is correct, but types are not being fetched during the initial list load.


#### The Problem
**Location:** `lib/services/api_service.dart:40-48`

```dart
final pokemonListItems = results
    .map(
      (item) => PokemonListItem(
        name: item['name'],
        url: item['url'],
        types: [],  //  TYPES ARRAY IS EMPTY!
      ),
    )
    .toList();
```

When the initial Pokemon list is fetched from the API, the `types` array is hardcoded as empty. The API endpoint `/pokemon?limit=20&offset=0` only returns Pokemon names and URLs, not their types.


## Issue #2: Favorites Persisting Across Different Users

### Severity: **HIGH** 

### Description
When a user favorites Pokemon and logs out, those favorites remain visible to the next user who logs in on the same device. This creates a data leakage issue where user preferences are shared across different accounts.
#### Problem 1: Global Storage Key
**Location:** `lib/services/local_storage.dart:6`

```dart
class LocalStorageService {
  static const String _favoritedPokemonKey = 'favorited_pokemon_cache';
  // This key is the SAME for ALL users!
```

The storage key is hardcoded and doesn't include any user identifier. All users on the same device share the same favorites cache.

#### Problem 2: No Cleanup on Logout
**Location:** `lib/services/auth.dart:72-78`

```dart
Future<void> signOut() async {
  try{
    return await _auth.signOut();  // Only signs out from Firebase
  }catch(e){
    debugPrint('Error signing out : $e');
  }
}
```

The sign-out process only clears Firebase authentication but doesn't clear local storage. Previous user's data remains on the device.

#### Problem 3: User ID Not Passed to Storage Service
**Locations:** `lib/controllers/pokemon_controller.dart`

- Line 141: `await _localStorage.getFavoritedPokemon();`
- Line 198: `await _localStorage.saveFavoritedPokemon(allFavorites.values.toList());`

The controller never passes the user ID to the storage service, making user-specific storage impossible with the current implementation.



### Affected Files
- `lib/services/local_storage.dart` (entire class)
- `lib/services/auth.dart` (signOut method)
- `lib/controllers/pokemon_controller.dart` (lines 139-142, 177-199, 214)



## Issue #3: Cache Service Missing Await on Async Operations

### Severity: **HIGH**  (Related to Issue #2)

### Description

**Location:** `lib/services/cache_service.dart:115-119`

The cache service's `getCachedPokemon()` method performs async removal operations without awaiting them, causing incomplete cache invalidation.

### Problem Code
```dart
Future<List<Pokemon>?> getCachedPokemon(String userId) async {
  final prefs = await SharedPreferences.getInstance();
  final cachedData = prefs.getString(_getUserKey(userId));

  if (cachedData != null) {
    final cacheTime = prefs.getInt(_getCacheTimeKey(userId));
    if (cacheTime != null) {
      final now = DateTime.now().millisecondsSinceEpoch;
      if (now - cacheTime > _cacheExpiry.inMilliseconds) {
        prefs.remove(_getUserKey(userId));        // ⚠️ NOT AWAITED
        prefs.remove(_getCacheTimeKey(userId));   // ⚠️ NOT AWAITED
        return null;
      }
    }
  }
  // ...
}
```

### Impact
- Cache entries may not be properly cleared
- Stale data could persist between user sessions
- **Contributes to Issue #2** (favorites persistence bug)
- Race conditions when cache is accessed immediately after invalidation


## Issue #4: Firebase Service Silent Failures

### Severity: **CRITICAL**  (Undermines Issue #2 Fix)

### Description

**Location:** `lib/services/firebase_service.dart:313-328`

The `_createUserDocument()` method catches Firestore errors but doesn't propagate them, leaving users without proper Firestore documents. This silently breaks the favorites persistence fix.

### Problem Code
```dart
Future<void> _createUserDocument(String uid) async {
  try {
    await _firestore.collection('users').doc(uid).set({
      'uid': uid,
      'favoritePokemonIds': [],
      'createdAt': FieldValue.serverTimestamp(),
    });
  } catch (e) {
    debugPrint('Error creating user document: $e');
    // ⚠️ ERROR NOT PROPAGATED - Caller doesn't know it failed!
  }
}
```

### Impact
- **Critical for Issue #2 Fix:** Without a Firestore document, favorites can't be persisted
- Users might silently fail to register properly
- App appears to work but favorites aren't saved to Firebase
- Network failures go unnoticed by calling code
- Debugging becomes extremely difficult

### Recommended Fix
Use the `Result` pattern already in the codebase:

```dart
Future<Result<void>> _createUserDocument(String uid) async {
  try {
    await _firestore.collection('users').doc(uid).set({
      'uid': uid,
      'favoritePokemonIds': [],
      'createdAt': FieldValue.serverTimestamp(),
    });
    return const Success(null);
  } catch (e, stackTrace) {
    debugPrint('Error creating user document: $e');
    return Failure(
      'Failed to create user document',
      exception: e as Exception,
      stackTrace: stackTrace,
    );
  }
}
```

Then update callers in `auth.dart`:

```dart
Future<User?> registerWithEmailAndPassword(String email, String password) async {
  try {
    // ... existing code ...

    final result = await _createUserDocument(user.uid);
    if (result.isFailure) {
      // Handle document creation failure
      await user.delete(); // Clean up Firebase Auth user
      debugPrint('Registration failed: ${result.errorOrNull}');
      return null;
    }

    return _userFromFirebaseUser(user);
  } catch(e) {
    debugPrint('Error registering : $e');
    return null;
  }
}
```




## Issue #6: Pokemon Controller Constructor Mismatch

### Severity: **HIGH** 

### Description

**Location:** `lib/controllers/pokemon_controller.dart:11, 73-75`

The controller declares `_apiService` as `final` but attempts to reassign it in the constructor. Additionally, `ApiService(http.Client())` doesn't match the parameterless constructor signature.

### Problem Code
```dart
class PokemonController extends ChangeNotifier {
  final ApiService _apiService = ApiService();  // Line 11

  // ... other code ...

  PokemonController() {
    _apiService = ApiService(http.Client());  // ⚠️ Can't reassign final field
  }                                            // ⚠️ Constructor mismatch
}
```





## Issue #7: Pokemon Model Missing Description Field Parsing

### Severity: **MEDIUM** 

### Description

**Location:** `lib/models/pokemon.dart:28-54`

The `Pokemon.fromJson()` factory constructor omits `description` field extraction, but `toJson()` includes it. This breaks round-trip serialization.

### Problem Code
```dart
factory Pokemon.fromJson(Map<String, dynamic> json) {
  try {
    final List<String> types = _extractTypes(json);
    final List<PokemonStat> stats = _extractStats(json);
    final String imageUrl = _extractImageUrl(json);
    final List<String> abilities = _extractAbilities(json);

    return Pokemon(
      id: json['id'] as int? ?? 0,
      name: json['name'] as String? ?? 'Unknown',
      imageUrl: imageUrl,
      types: types,
      stats: stats,
      height: json['height'] as int? ?? 0,
      weight: json['weight'] as int? ?? 0,
      abilities: abilities,
      // ⚠️ Missing: description: json['description'] as String?
    );
  } catch (e) {
    throw FormatException('Failed to parse Pokemon from JSON: $e');
  }
}

Map<String, dynamic> toJson() {
  return {
    'id': id,
    'name': name,
    'imageUrl': imageUrl,
    'types': types,
    'stats': stats.map((s) => s.toJson()).toList(),
    'height': height,
    'weight': weight,
    'description': description,  // ⚠️ Included in toJson()
    'abilities': abilities,
  };
}
```

### Impact
- Data loss when caching/reloading Pokemon
- Description is lost after serialize/deserialize cycle
- Cache invalidation issues



## Issue #8: API Service Missing Timeouts and Error Handling

### Severity: **MEDIUM** 

### Description

**Location:** `lib/services/api_service.dart:33, 91`

HTTP requests don't have timeouts and catch blocks silently swallow errors without logging.

### Problems
1. **No Timeout (Line 33):**
```dart
final response = await _client.get(Uri.parse('$_baseUrl/pokemon'));
//  Can hang indefinitely on slow networks
```

2. **Silent Error Swallowing (Line 91):**
```dart
} catch (e) {
  //  Empty catch block - errors disappear
}
```

### Impact
- App can freeze on slow networks
- Debugging becomes impossible when errors are swallowed
- Poor user experience with no timeout feedback


---




---

itumelengseema committed:
refactor: clean up import statement in main.dart

itumelengseema committed:
refactor: implement dotenv for environment variables and enhance error handling in API service

itumelengseema committed:
refactor: streamline API service initialization and clean up unused imports

itumelengseema committed:
fix: resolve favorites carry-over bug and add empty state UI

- Filter favorites cache to only show Pokemon belonging to current user
- Clear favorites cache when user changes or signs out
- Add 'No Favorites Found' empty state UI with helpful message
- Prevent mysterious favorites from appearing for new users

itumelengseema committed:
refactor: add .env file for environment variables and update asset references

Joashp01 reviewed (approved):
approved
